pkg/updatestrategy/drain_test.go

```diff
@@ -0,0 +1,503 @@
+		}
+
+		ctx, cancel := context.WithCancel(context.Background())
+
+		go func() {
+			// wait until evict() unblocks us
+			blockHelper.Wait()
+			cancel()
+			// unblock evict()
+			blockEviction.Done()
+		}()
+
+		err := mgr.TerminateNode(ctx, &Node{Name: node.Name}, false)
+
+		evictCountFinal := atomic.LoadInt32(&evictCount)
+		assert.Equal(t, int32(2), evictCountFinal)
+		assert.EqualValues(t, err, context.Canceled)
+	}
+
+	// cancel during slow eviction
+	{
+		var deleteCount int32
+		evictPod = evictPodFailPDB
+
+		mgr := &KubernetesNodePoolManager{
+			logger:  logger,
+			kube:    setupMockKubernetes(t, []*v1.Node{node}, pods, nil),
+			backend: backend,
+			drainConfig: &DrainConfig{
+				ForceEvictionInterval: time.Minute,
+				PollInterval:          time.Second,
+			},
+		}
+
+		ctx, cancel := context.WithCancel(context.Background())
+		deletePod = func(client kubernetes.Interface, logger *log.Entry, pod v1.Pod) error {
+			atomic.AddInt32(&deleteCount, 1)
+			cancel()
+			return nil
+		}
+		err := mgr.TerminateNode(ctx, &Node{Name: node.Name}, false)
+
+		deleteCountFinal := atomic.LoadInt32(&deleteCount)
+		assert.EqualValues(t, int32(1), deleteCountFinal)
+		assert.EqualValues(t, err, context.Canceled)
+	}
+}
+
+var testDrainConfig = &DrainConfig{
+	ForceEvictionGracePeriod:       10 * time.Second,
+	MinPodLifetime:                 20 * time.Second,
+	MinHealthyPDBSiblingLifetime:   30 * time.Second,
+	MinUnhealthyPDBSiblingLifetime: 40 * time.Second,
+	ForceEvictionInterval:          5 * time.Second,
+}
+
+func forceTerminationAllowed(t *testing.T, pods []*v1.Pod, pdbs []*policy.PodDisruptionBudget, now, drainStart, lastForcedTermination time.Time) (bool, error) {
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: testNodeName,
+		},
+	}
+
+	backend := &mockProviderNodePoolsBackend{
+		nodePool: &NodePool{
+			Min:        1,
+			Max:        1,
+			Current:    1,
+			Desired:    1,
+			Generation: 1,
+			Nodes: []*Node{
+				{ProviderID: "provider-id"},
+			},
+		},
+	}
+
+	mgr := &KubernetesNodePoolManager{
+		logger:      log.WithField("test", true),
+		kube:        setupMockKubernetes(t, []*v1.Node{node}, pods, pdbs),
+		backend:     backend,
+		drainConfig: testDrainConfig,
+	}
+
+	return mgr.forceTerminationAllowed(*pods[0], now, drainStart, lastForcedTermination)
+}
+
+var zeroTime = time.Unix(0, 0)
+
+func pod(name string, creationTimestamp time.Time, healthy bool) *v1.Pod {
+	status := v1.ConditionFalse
+	if healthy {
+		status = v1.ConditionTrue
+	}
+
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              name,
+			Namespace:         "default",
+			CreationTimestamp: metav1.Time{Time: creationTimestamp},
+			Labels:            map[string]string{"app": "foo"},
+		},
+		Spec: v1.PodSpec{
+			NodeName: testNodeName,
+		},
+		Status: v1.PodStatus{
+			Conditions: []v1.PodCondition{{Type: v1.PodReady, Status: status}},
+		},
+	}
+}
+
+func TestForceTerminationGracePeriod(t *testing.T) {
```

Comment:
You can put this + the next 2 tests together and build a so called "table driven test".

Example:
``` 
func TestSkipperCustomRoutes(t *testing.T) {
	for _, ti := range []struct {
		msg            string
                pods             []*v1.Pod
		expected    bool
	}{{
		msg: "test1",
		pods: []*v1.Pod{pod("foo-0", zeroTime, true)},
		expected: true,
        }, {
		msg: "test2",
		pods: []*v1.Pod{pod("foo-0", zeroTime, true)},
		expected: false,

        }, {
		msg: "test3",
                  ...
	}} {
		t.Run(ti.msg, func(t *testing.T) {
			.... result = ..
                        if ti.expected != result {
                            ti.Errorf("...")
                        }  
		})
	}
}
```

Reply:
I think it's easier to deal with separate test cases as long as there's not a lot of duplicate code. The error messages you get are more clear and you don't have to manually reimplement the logic.